Convert PluginPropertiesExtension Groovy to Java

rhamedy opened the pull request:
Contributes to #34459

Converted
- PluginPropertiesExtension.groovy -> java 
- Deleted the groovy files

rhamedy committed:
Convert PluginPropertiesExtension & PluginPropertiesTask Groovy to Java

rhamedy commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,125 @@
+        substitutions.put("elasticsearchVersion", Version.fromString(VersionProperties.getElasticsearch()).toString());
+
+        //TODO (Raf): No getter for project.getTargetCompatibility()? What's the alternative option for below line?
+        substitutions.put("javaVersion", String.valueOf(this.getProject().getVersion()));
```
I could not find a getter for `project.getTargetCompatibility()`? Would appreciate feedback for `javaVersion` if `this.getProject().getVersion()` is not correct alternative.

atorok replied:
I think that's a java plugin conventions, you should be able to access it from Java in the same way we access source sets in existing code.

rhamedy commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,125 @@
+        substitutions.put("classname", this.extension.getClassname());
+
+        //TODO (Raf): Should we worry about possible NPE on getExtendedPlugins() here?
+        substitutions.put("extendedPlugins", this.extension.getExtendedPlugins().stream().collect(Collectors.joining(",")));
```
Will the `.getExtendedPlugins()` ever return `NULL`? If yes, then I will add a null-check otherwise the current `implementation` should suffice.

atorok replied:
I think it's fine, as long as we check it in the extension.

rhamedy commented:
Tried to convert the `BuildPluginBuild.groovy` to `java` and include it in this PR however, I hit a few roadblocks relating to  
- Replacing `project.pluginProperties.extension.name equivalent`with `project.getTasks().getByName("pluginProperties").getName()`? 
- How to access `module` in `project.integTestCluster.module(project)`? 
- No `publishing` property on `project` for `project.publishing.publications.nebula(MavenPublication).artifactId`
- A few others like this 

I have paused working on `BuildPluginBuild.groovy` because of one too many question marks. Not sure if this PR can go through independent of `BuildPluginBuild` refactor?

elasticmachine commented:
Pinging @elastic/es-core-infra

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,173 @@
+        Object version = project.getVersion();
+
+        this.name = project.getName();
+        this.version = (version != null && version instanceof String)? String.valueOf(version) : null;
```
We don't check against null lather on, maybe we shouldn't allow it.

rhamedy replied:
Replaced it with set the version if `if(version != null && version instanceof String)` returns true

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,173 @@
+        this.project = project;
+    }
+
+    @Input
```
These annotations are not read on extensions, only on tasks, they were there in Groovy by error, please remove them.

rhamedy replied:
Removing all annotations i.e. `@Input`, `@Output`, `@InputFile` etc.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,173 @@
+        return project;
+    }
+
+    public void setProject(Project project) {
```
There's no need for this setter, we can set this in the constructor and mark it `final`.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,125 @@
+
+        File templateFile = new File(project.getBuildDir(), "templates/" + this.descriptorOutput.getName());
+
+        Task copyPluginPropertiesTemplate = project.getTasks().create("copyPluginPropertiesTemplate")
```
We don't want to configure in the constructor for anything Java related. 
Can you please rewrite this to use `project.copy` in the `@TaskAction`.

The validations on the `afterEvaluate` should go into `getExtension` since we want these to fail even if the task actions ( i.e. the task itself ) are not executed.

rhamedy replied:
@atorok I would need some feedback around what I have done for 
```
We don't want to configure in the constructor for anything Java related.
Can you please rewrite this to use project.copy in the @TaskAction.
``` 
I wrote something as follow inside `@TaskAction` 
```
project.copy((unused) -> {
       from("/" + this.descriptorOutput.getName());
       into(templateFile.toPath());
 });
```
but, what about the line with `dependsOn`? We don't have a reference to task. First, time I am working with `Gradle` hence, why the struggle I guess.

atorok replied:
You won't need the `dependsOn` ( it's `this.dependsOn` btw ). 
And you need to ditch the `extends Copy` just  extend `DefaultTask`. 
`project.copy` behaves like the `Copy` task except it runs when you call it, it's not a task, but it's configured in the same way. The `unused` variable is what you use to configure it. What you wrote above is a nop, a copy that does nothing but configure the task it's called in as  then  `from` refers to configuration on the task, that has no effect anyway because we are already in the execution phase when `@TaskAction` runs.  Since you no longer extend the `Copy` task, you'll have to use another `project.copy` for that functionality.  There will be no `project.tasks.create('copyPluginPropertiesTemplate')` either, as that's going to be replaced with  a `project.copy` too.

rhamedy replied:
I pushed some changes along with unit tests 👍

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,173 @@
+    private String classname;
+
+    /** Other plugins this plugin extends through SPI */
+    private List<String> extendedPlugins = new ArrayList<>();
```
this should be marked `final`

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,173 @@
+
+    @Input
+    public List<String> getExtendedPlugins() {
+        return extendedPlugins;
```
It's better to return an `Collections.unmodifiableList` so callers can't change the list.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,173 @@
+    }
+
+    public void setExtendedPlugins(List<String> extendedPlugins) {
+        this.extendedPlugins = extendedPlugins;
```
We should add a null check on the arguments.

rhamedy replied:
If we marked it `final` as per above comment then this setter goes away. No?

atorok replied:
I was considering making it final, going away with the setter and having an `extendedPlugin` method to add them, but I didn't want to impose that on this PR, the comment on final did sneak in trough, so sorry for the confusion. 
If we don't do the above, we'll need to keep the setter and the field can't be final, so just add the null check and we are fine.

rhamedy replied:
No worries. Just to make sure I am not confused. I will remote the `final` keyword from the `extendedPlugins` array list field and add a `setter` with a null check to only set the `extendedPlugins` if its not null. This way it will either have extended plugins or be empty as per declaration 👍

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,125 @@
+        this.extension = extension;
+    }
+
+    @InputFile
```
This should be `@OutputFile`

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,125 @@
+        return extension;
+    }
+
+    public void setExtension(PluginPropertiesExtension extension) {
```
I don't think we should have a setter for it, the task should discover it from project.

alpar-t reviewed (changes requested):
Thank you for your contribution @rhamedy !
We prefer small PRs so this is even better than having the plugin converted too. 
I left a few comments. Also would be kind enough and add tests for these classes ? 
We want all the java code to have tests.

rhamedy committed:
Addressed review comments

rhamedy commented:
@atorok thank you for the review. I have addressed them in my recent `commit` as well as replied to some of your comments that need a reply.

I would definitely write tests! Is there a `test class` that I could use as a guide? I need a little sense of direction on how write tests 🙂

rhamedy commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -60,17 +60,20 @@
         this.name = project.getName();
-        this.version = (version != null && version instanceof String)? String.valueOf(version) : null;
+
+        if(version != null && version instanceof String) {
```
It would still be `null` if the condition is false 😕

atorok commented:
Thanks @rhamedy  for the quick update.  For tests I think you should look at `GradleUnitTestCase` and other classes that implement it. 

@elasticmachine test this please

rhamedy committed:
Refactored the code with regards to code review comments

rhamedy committed:
Added plugin-descriptor.proerties along with unit tests for new java classes

rhamedy commented:
@atorok I have addressed the code review comments as well as added some tests. I could not find a class that I could use as baseline to write tests probably because the tests are different in other classes and this is plugin. 

The test ensures tha the `plugin-descriptor.properties` file is picked up from classpath and it's substituted with `props` from `extension` and it's saved in the project `build directory` under `templates`. 

Please let me know if any changes to be made :+1:

atorok commented:
@elasticmachine test this please

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,164 @@
+        this.name = project.getName();
+
+        if(version != null && version instanceof String) {
+            this.version = String.valueOf(version);
```
version should always be a string. This conditional should not be necessary.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,164 @@
+        this.classname = classname;
+    }
+
+    public Collection<String> getExtendedPlugins() {
```
extended plugins was a List previously, I think it should remain that, not be reduced to a Collection?

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,164 @@
+    }
+
+    public void setExtendedPlugins(List<String> extendedPlugins) {
+        if(extendedPlugins != null) {
```
Why this leniency? If we set extendedPlugins, then it should be set. This would result in very difficult to debug behavior if someone extendedPlugins was set once before.

rhamedy replied:
Please check this review comment by @atorok  https://github.com/elastic/elasticsearch/pull/39605#discussion_r262972790 (recommended adding a null-check)

rjernst replied:
I think a null check is incorrect. As I noted above, it means it is "set once", but silently skips future sets.

rhamedy replied:
Make sense! I will drop the null-check.

atorok replied:
This is not what I had in mind with the null check. 
We should fail fast if the extenedPlugins is being set to a `null`. 
It's sometimes easy to set it to something in a build script that can return a null, we should should fail the build right away by throwing an exception as it eases debugging and makes it obvious where the null came from.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,108 @@
+        File templateFile = new File(project.getBuildDir(), "templates");
+        templateFile.getParentFile().mkdirs();
+
+        project.copy((cp) -> {
```
nit: the parens are not necessary around `cp`

rhamedy replied:
Sorry! I do not understand the comment :neutral_face:

rjernst replied:
You can remove the parenthesis around cp, ie `(cp) -> {` becomes `cp -> {`

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,108 @@
+        templateFile.getParentFile().mkdirs();
+
+        project.copy((cp) -> {
+            cp.from(getClass().getClassLoader().getResource(this.descriptorOutput.getName())).into(templateFile);
```
This copying needs to be using a stream like it did before (using getResourceAsStream and copying in a doLast). The zip filesystem java puts over jars has issues. IIRC I had issues with this failing when build-tools is bundled into a jar (ie for an external plugin).

rhamedy replied:
This review comment https://github.com/elastic/elasticsearch/pull/39605#discussion_r263697609 by @atorok pointed me in the direction of using `projct.copy` and I thought perhaps I should make use of `from` and `into` inside the `project.copy`. 

I can replace this with with `getResourceAsStream` but, want to be sure that the upcoming changes are not conflicting with earlier review comments. 

Why do we need the `doLast` part?

rjernst replied:
AFAIK, there is no CopySpec method that takes in input stream. So in order to use getResourseAsStream, we need to actually copy the bytes from the stream, to a file (as we did before). Thus we need some arbitrary code to run, in a doLast.

rhamedy replied:
@rjernst How does the `doLast` block get triggered? I have the follow task action 
```
@TaskAction
    public void configurePluginPropertiesTask() {
        Project project = this.getProject();
        
        File templateFile = new File(project.getBuildDir(), "templates/" + descriptorOutput.getName());
        templateFile.getParentFile().mkdirs();

         doLast(unused -> {
                InputStream resourceTemplate = this.getClass().getResourceAsStream("/" + descriptorOutput.getName());
                templateFile.getParentFile().mkdirs();
                try {
                    Files.write(templateFile.toPath(), resourceTemplate.readAllBytes());
                } catch (IOException e) {
                    throw new RuntimeException(e);
                }
            });
            
        project.copy(copyArg -> {
            copyArg.from(templateFile.getParentFile()).include(descriptorOutput.getName())
                .into(descriptorOutput.getParentFile()).expand(this.generateSubstitutions());
            this.getInputs().properties(this.generateSubstitutions());
        });
    }
``` 
and calling `pluginPropertiesTask.configurePluginPropertiesTask();` ignores the `doLast` 😕

rjernst replied:
A couple things here:
* The name of this method is confusing, since it is not configuration; it is the actual action for the task
* Don't use readAllBytes on the input stream. Instead call `Files.copy(resourceTemplate, templateFIle.toPath())`.
* The doLast shouldn't be on the pluginProperties task. Instead, it was previously on an ancillary task whose sole purpose was to copy the template from the build-tools jar into the plugin project. I see now Alpar asked to not do that anymore. That is fine. In this case, the contents of the doLast can simply be removed. The contents should be run just before the project.copy you have.

atorok replied:
The `@TaskAction` is in the same realm as `doLast` in that each Gradle task has a list of  actions to execute. `doFirst` and `doLast` are ways to add to that list, but like Ryan points it out are not necessary here, or in custom tasks in general as you are implementing the main task action and can freely move code around.

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,108 @@
+        substitutions.put("name", this.extension.getName());
+        substitutions.put("description", this.extension.getDescription());
+        substitutions.put("version", this.extension.getVersion());
+        substitutions.put("elasticsearchVersion", Version.fromString(VersionProperties.getElasticsearch()).toString());
```
I'm not sure why I did this before, but converting to a Version object should be unnecessary; we already have the version as a string.

rhamedy replied:
I will check the project for how this is done :+1:

atorok replied:
I do think this is necessary, granted we should express it better.  The `Version` class strips away snapshot and qualifier so `8.0.0-SNAPSHOT` becomes `8.0.0` and `7.0.0-rc1` becomes `7.0.0`

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,164 @@
+        return Collections.unmodifiableCollection(extendedPlugins);
+    }
+
+    public boolean isHasNativeController() {
```
we can just use `has` here, no need for `isHas`

rjernst commented on buildSrc/src/test/resources/plugin-descriptor.properties:
```diff
@@ -0,0 +1,45 @@
+# Elasticsearch plugin descriptor file
```
Why is a duplicate file being added under test?

rhamedy replied:
Since I am using the `getClass().xxxxxx` in the test it is looking into `src/test/resource` for `plugin-descriptor.properties` in order to copy it over to `templates` and then perform `property substitution`. 

See https://github.com/elastic/elasticsearch/pull/39605/files#diff-a9a997ce88a166a144b1881bc9783ce6R66

rjernst replied:
This duplicate is not necessary.  Also, isn't the test trying to load the *output* file, not the template?

rhamedy replied:
You were right! The most recent helped me understand how the plugin works and I have tweaked the test as well.

rjernst reviewed (changes requested):
Thanks for taking this on @rhamedy. I left a few comments.

rhamedy commented:
@rjernst and @atorok thank you for the review. Replied to the comments. Do you have any feedback on unit tests? It would be nice to tackle the review changes along with tests all at once.

rhamedy committed:
Removed unnecessary properties file

rhamedy committed:
Changed the implmentation and addressed code review

rhamedy committed:
Fixed the tests and verified its working

rhamedy commented:
@rjernst @atorok I pushed up my new changes taking into account the code review comments. I have removed the unecessary `test/resources/plugin-descriptor.properties` file and after throwing away the `doLast` everything works magically and according to my tests the `templates` and `generated-resources` directories have correct content 👍

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,154 @@
+
+    public PluginPropertiesExtension(Project project) {
+        this.name = project.getName();
+        this.version = project.getVersion().toString();
```
I just realized this now, this is a problem present in the old code as  well. 
The way we assign this default makes it prone to ordering. 
If the extension is created before `project.version` is assigned we won't pick it up. 
We should have a null check in the getter instead  and return`project.version` if `version` is `null`
This is fine also in a subsequent PR.

rhamedy replied:
I can update the `PluginPropertiesExtension.getVersion()` to return `project.getVersion().toString()` if `version` is null :+1:

rjernst replied:
It is not possible for version to be null. VersionProperties statically loads the version. BuildPlugin   is initialized before any PluginPropertiesExtension, since it is applied by PluginBuildPlugin, which extends BuildPlugin. Adding null checks all over for version will only make the code more complicated to understand with no actual additional safety.

atorok replied:
We do this early on in the root level `build.gradle` 
```
allprojects {
  group = 'org.elasticsearch'
  version = VersionProperties.elasticsearch
  description = "Elasticsearch subproject ${project.path}"
}
```
So indeed it's not possible for `project.version` to be null in _our_ build, 
but all bets are off for external plugins, there's no guarantee that it won't be null. 
We might not want to be lenient here and throw an exception if the version is null, but I still think we should do it rather than leave it be an NPE. 
Alternatively, we could hard code this to `VersionProperties.elasticsearch` as we only test building plugins where the version of build-tools matches that of elasticsearch. 
External plugins can then have whatever project.version they like. 
What do you think of this approach @rjernst

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,115 @@
+
+    public PluginPropertiesTask() {
+        Project project = this.getProject();
+        this.extension = project.getExtensions().create("esplugin", PluginPropertiesExtension.class, project);
```
sorry I missed this in my previous review,
Creating the extension here  is part of the configuration we want to avoid doing as part of the constructor.
The proper place for it is in the `PluginBuilderPlugin`. We can then reference it here ( using get by name instead of create ) and we must also declare it as an `@Input` as we want the task to be re-run if any of it changes.

rhamedy replied:
Can you help me with understanding what is `PluginBuilderPlugin` I could not find a reference to it anywhere? :confused:

atorok replied:
Sorry, I made a typo in the name the class is called `PluginBuildPlugin`

rhamedy replied:
I might need help with this one. I don't know how to modify the `PluginBuildPlugin.groovy` and I tried adding the the `plugin` as follow (using create, add, etc.)  

```project.extensions.create('esplugin', PluginPropertiesExtension, project)``` 

And seems like it adds it but, then in the `PluginPropertiesTask` inside the constructor 

```this.extension = project.getExtensions().getByType(PluginPropertiesExtension.class);``` 
and 
```this.extension = (PluginPropertiesExtension) project.getExtensions().getByName('esplugin');```

Fails complaining that it cannot find the plugin. 

```
Caused by: org.gradle.api.UnknownDomainObjectException: Extension of type 'PluginPropertiesExtension' does not exist. Currently registered extension types: [ExtraPropertiesExtension, DefaultArtifactPublicationSet, ReportingExtension, SourceSetContainer, JavaPluginExtension]
``` 
The above is when I try to retrieve the plugin via `getByType` in the `PluginPropertiesTask` and when I try to retrieve it with `getByName` then error is as follow 

```
org.gradle.api.UnknownDomainObjectException: Extension with name 'esplugin' does not exist. Currently registered extension names: [ext, defaultArtifacts, reporting, sourceSets, java]
```
It is almost like `accessible` in `PluginBuildPlugin` (because when I try it again, it complains that it's already registered) but, not accessible in `PluginPropertiesTask`.

atorok replied:
Make sure that the extension is registered _before_ the task is created. 
That seems obvious but sometimes it's easy to overlook ordering with Gradle. 
Other than that what you are doing looks right.
Tasks also have extensions, in a different namespace, but you are correctly trying to look it up from the project.

atorok replied:
@rhamedy thanks for continuing to pursue this!  If you can't get it to work feel free to push the code and I'll take a look.

rhamedy replied:
I have done the above refactoring! Please see the recent changes.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,115 @@
+        File templateFile = new File(project.getBuildDir(), "templates/" + descriptorOutput.getName());
+        templateFile.getParentFile().mkdirs();
+
+        InputStream resourceTemplate = this.getClass().getResourceAsStream("/" + descriptorOutput.getName());
```
the input stream is never closed, we should use try-with-resource here.

rhamedy replied:
Good catch! Will use the `try-with-resource` here :+1:

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,115 @@
+        project.copy(copyArg -> {
+            copyArg.from(templateFile.getParentFile()).include(descriptorOutput.getName())
+                .into(descriptorOutput.getParentFile()).expand(this.generateSubstitutions());
+            this.getInputs().properties(this.generateSubstitutions());
```
This is what I was referring to in my previous comment w.r.t marking the extension as an `@Input`.
It's add to do it part of a copy action configuration, the clean way is to make `generateSubstitutions` public and mark it wiht `@Input` this line is not needed any more then.

rhamedy replied:
Please check the recent changes.

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,115 @@
+        return substitutions;
+    }
+
+    @OutputFile
```
This annotation is incorrect here.
This could move into `generateSubstitutions` I think that's a proper place to do those checks.

rhamedy replied:
I will remote the `OutputFile` annotation. Why are we moving these `if-checks`?

atorok replied:
Without the annotations the method will never be called.

atorok commented:
@elasticmachine  test this please

rhamedy committed:
Register extension in PluginBuildPlugin, Refactore PluginPropertiesTask, update the tests

rhamedy committed:
Wrap the inputstream in try-with-resource

rhamedy commented:
@atorok @rjernst I have factored the changes requested and pushed the changes up. The tests that I have added are working 👍 however when running the command `./gradlw check` I get the following error

```
> Task :modules:aggs-matrix-stats:pluginProperties FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':modules:aggs-matrix-stats:pluginProperties'.
> java.nio.file.FileAlreadyExistsException: /Users/raf/workspace/opensource-contributions/elastic/elasticsearch/modules/aggs-matrix-stats/build/templates/plugin-descriptor.properties
``` 
I double checked the `modules:aggs-matrix-stats` and unlike the exception the file `modules/aggs-matrix-stats/build/templates/plugin-descriptor.properties` is `present 🤔

atorok commented:
@elasticmachine test this please

rhamedy commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:
```diff
@@ -42,6 +42,10 @@ public class PluginBuildPlugin extends BuildPlugin {
         super.apply(project)
+
+        project.extensions.create('esplugin', PluginPropertiesExtension, project)
+        project.pluginManager.apply('elasticsearch.esplugin')
```
You mean `line 47`? or the entire change?

rjernst replied:
Yes, line 47. `PluginBuildPlugin` *is* `elasticsearch.esplugin`, so it has already been applied.

rhamedy replied:
Done! 👍

rhamedy commented:
@atorok @rjernst please let me know if any further changes need to be made. I will clean up (squash) the commits after the PR is in mergeable state 🙂

alpar-t reviewed (approved):
LGTM no need to squash we do that on merge anyhow.

rjernst commented:
@elasticmachine test this please

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,112 @@
+        }
+
+        project.copy(copyArg -> {
+            copyArg.from(templateFile.getParentFile()).include(descriptorOutput.getName())
```
nit: please separate these chained calls to make it more clear all the operations that are happening

rhamedy replied:
I will separate these with each in its own line 👍

rjernst commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,112 @@
+        templateFile.getParentFile().mkdirs();
+
+        try (InputStream resourceTemplate = this.getClass().getResourceAsStream("/" + descriptorOutput.getName())) {
+            Files.copy(resourceTemplate, templateFile.toPath());
```
I think you need StandardCopyOption.REPLACE_EXISTING since the file may have already been copied from a previous run. This would cause the file exists error you saw.

rhamedy replied:
That make sense. Thanks for clarifying it.

rhamedy committed:
Use StandardCopyOption.REPLACE_EXISTING when File.copy, unchain the method calls

atorok commented:
@elasticmachine test this please

rhamedy commented:
@atorok not sure what's conflicting? Should I resolve it?

atorok commented:
@rhamedy  yes, we do need to solve it. @mark-vieira replaced one of the custom tasks in this PR with a default task because most of the config moved to the plugin but there are I think still a few classes that we could keep from this conversion.

rhamedy commented:
Done! Curious what's blocking merge of changes in this PR? 🤔

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesTask.java:
```diff
@@ -0,0 +1,115 @@
+/**
+ * Creates a plugin descriptor.
+ */
+public class PluginPropertiesTask extends DefaultTask {
```
We don't want this class back now, it will go unused now as the functionality was rolled into the plugin in the mean time. 
Converting `PluginPropertiesExtension` to java is what we need to keep from this PR.

rhamedy replied:
@atorok trying to run the test or even `./gradlew build` or `check` or `etc` results in 
```
Caused by: groovy.lang.GroovyRuntimeException: Cannot get the value of write-only property 'hasNativeController' for object of type org.elasticsearch.gradle.plugin.PluginPropertiesExtension.
```  
I have changed the `hasNativeController` to `getHasNativeController` and the exception disappears. @rjernst had suggested this change (to keep it `hasNativeController`) however, since rebase this issue come up. 

For now, I have changed it to `getHasNativeController`. I have tried running the tests as follow 

```
./gradlew :build-tools:test "-Dtests.class=org.elasticsearch.gradle.plugin.PluginProprtiesExtensionTests"
``` 
but, since rebase the output is different. There is no signs of tests being run 😕

atorok replied:
we switched to running tests with the Gradle test runner, so the syntax to filter tests to be ran is now `./gradlew :build-tools:test --tests=*PluginProprtiesExtensionTests`

rhamedy replied:
Thanks. The docs doesn't reflect the change 🤔

rhamedy committed:
Delete previously converted PluginPropertiesTask and it's tests, rename hasNativeController to getHasNativeController

atorok commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/plugin/PluginBuildPlugin.groovy:
```diff
@@ -52,6 +52,8 @@ class PluginBuildPlugin extends BuildPlugin {
         super.apply(project)
 
         PluginPropertiesExtension extension = project.extensions.create(PLUGIN_EXTENSION_NAME, PluginPropertiesExtension, project)
+        project.pluginManager.apply('elasticsearch.esplugin')
```
No need to apply here, the plugin is already applied on line 52

atorok commented on buildSrc/src/main/java/org/elasticsearch/gradle/plugin/PluginPropertiesExtension.java:
```diff
@@ -0,0 +1,154 @@
+        return this.extendedPlugins;
+    }
+
+    public boolean getHasNativeController() {
```
The name is inconsistent with other methods that return boolean and are prefixed with `is`

rhamedy replied:
Good point! I was being lazy 😞

atorok commented:
@elasticmachine test this please

rhamedy committed:
removed apply plugin and updated hasNativeController getter name

rhamedy commented:
@atorok pushed my changes! 👍

atorok commented:
@elasticmachine test this please

atorok commented:
@rhamedy looks like there are some failing tests

rhamedy committed:
set license and notice file on extensions extra properties

rhamedy commented:
@atorok that recent changes fix the test failures 👍

atorok commented:
@elasticmachine test this please

atorok commented:
Sorry for the delay @rhamedy I'll see to merge this as soon as the PR checks pass

alpar-t committed:
Merge master and make version lazy

Merge remote-tracking branch 'origin/master' into plugin-properties-task-tojava
When the version is not set, use the project version bu do so lazily so
we don't snapshot it at construction time

rhamedy commented:
That would be great! 🙂
